Amazon.Ion.ObjectMapper/IonObjectSerializer.cs

```diff
@@ -22,47 +22,75 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public object Deserialize(IIonReader reader)
         {
-            var targetObject = options.ObjectFactory.Create(options, reader, targetType);
-            reader.StepIn();
-
-            IonType ionType;
-            while ((ionType = reader.MoveNext()) != IonType.None)
+            if (reader == null)
+            {
+                throw new InvalidOperationException("Cannot deserialize with a null Ion reader");
+            }
+            
+            try
             {
-                var property = FindProperty(reader.CurrentFieldName);
-                FieldInfo field;
-                if (property != null)
+                reader.StepIn();
+                
+                var ionConstructors = targetType.GetConstructors().Where(IsIonConstructor);
+                switch (ionConstructors.Count())
                 {
-                    var deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
-                    
-                    if (options.IgnoreDefaults && deserialized == default)
-                    {
-                        continue;
-                    }
-                    
-                    property.SetValue(targetObject, deserialized);
+                    case 1:
+                        return InvokeIonConstructor(ionConstructors.First(), reader);
```

Comment:
So if you've indicated there's a constructor, then no other setters are called? Is that expected?

Reply:
The idea behind IonConstructor is to just look for the constructor's parameters in the Ion and call the constructor with those parameters rather than the usual deserialization logic.